orchagent/watermarkorch.cpp

```diff
@@ -170,16 +207,18 @@ void WatermarkOrch::doTask(SelectableTimer &timer)
 
```

Comment:
For line 198 to 206, how can you be so sure that periodic timers kicks in first before user issues the clear watermark through cli?  That m_pg_ids, m_unicast_queue_ids, and m_multicast_queue_ids is already populated for the latter?

Reply:
The idea of introducing m_pg_ids, m_unicast_queue_ids, and m_multicast_queue_ids , I guess, is to have a cached copy of the oids at hand to remove the need of further queries into obtaining these oids. The only opportunity/condition to update this cached copy is when a watermark clear command is issued by timer and the doTask() handler finds that the vector is empty. So we assume here that when this update opportunity comes, the pg, uni_queue, and multi_queue oids are generated complete and will not be subject to any further changes.

The sources of the oids comes from the COUNTERS_DB *MAP table, which is set when FlexCounterOrch receives any signal on "FLEX_COUNTER_STATUS" field to call the generate*Map() function. This signal in the current orchagent initialization sequence, I guess, comes for granted after the port initialization is done, with all pg and queue oids being settled. So we will get a list of pg and queue oids complete and ready at the time the cached oid copy still hits the emptiness condition.

Reply:
The main idea of having m_pg_ids, m_unicast_queue_ids, m_multicast_queue_ids is to avoid querying them every time, performing deserialization of oids.  This things are not expected to change.